private/buf/bufsync/syncer.go

```diff
@@ -94,508 +75,655 @@ func newSyncer(
-	for _, branch := range branchesToSync {
-		if branch == gitDefaultBranch {
-			continue // default branch will be injected manually
-		}
-		sortedBranchesForSync = append(sortedBranchesForSync, branch)
-	}
-	sort.Strings(sortedBranchesForSync)
-	s.sortedBranchesForSync = append([]string{gitDefaultBranch}, sortedBranchesForSync...) // default first, then the rest A-Z
-	for _, moduleDir := range s.sortedModulesDirsForSync {
-		s.modulesDirsToBranchesToIdentities[moduleDir] = make(map[string]bufmoduleref.ModuleIdentity)
-		for _, branch := range s.sortedBranchesForSync {
-			s.modulesDirsToBranchesToIdentities[moduleDir][branch] = nil
-		}
+	// Load all tagged commits ahead of time.
+	commitHashToTags := make(map[string][]string)
+	if err := s.repo.ForEachTag(func(tag string, commitHash git.Hash) error {
+		commitHashToTags[commitHash.Hex()] = append(commitHashToTags[commitHash.Hex()], tag)
+		return nil
+	}); err != nil {
+		return nil, nil, err
 	}
-	// (3) Prepare module targets for all module directories and branches.
-	allModulesIdentitiesForSync := make(map[string]bufmoduleref.ModuleIdentity) // moduleIdentityString:moduleIdentity
+	var (
+		moduleBranches []ModuleBranch
+		// All tagged commits for a module identity. The commits in this set are either already synced or
+		// will be synced in at least one ModuleBranch.
+		taggedCommitsToSyncForModuleIdentity = make(map[bufmoduleref.ModuleIdentity]map[git.Commit][]string)
+	)
+	// Walk branches and collect ModuleBranches and tagged commits to sync. The order of branches here
+	// doesn't matter, everything will be ordered at the end.
 	for _, branch := range branchesToSync {
 		headCommit, err := s.repo.HEADCommit(
 			git.HEADCommitWithBranch(branch),
 			git.HEADCommitWithRemote(s.gitRemoteName),
 		)
 		if err != nil {
-			return fmt.Errorf("reading head commit for branch %s: %w", branch, err)
+			return nil, nil, fmt.Errorf("reading head commit for branch %s: %w", branch, err)
 		}
+		moduleDirsFoundForModuleIdentity := make(map[string][]string) // moduleIdentity:[]moduleDir
+		// Walk all module dirs for a branch. The order of module dirs here doesn't matter, everything will
+		// be ordered at the end.
 		for moduleDir, identityOverride := range s.modulesDirsToIdentityOverrideForSync {
+			// Determine the module identity to sync this moduleDir with, in this branch.
+			// If the user provided an override, prefer it always. Otherwise use the identity of the
+			// module at HEAD.
+			builtModule, readErr := s.readModuleAt(ctx, headCommit, moduleDir)
+			if readErr != nil {
+				// Module read failed. Either this is a transient error, or the module is invalid, either
+				// way, the reason doesn't really matter, syncing this branch should fail sync. There may
+				// be other commits behind HEAD that are valid to sync, but for now nothing on this branch
+				// will be synced.
+				return nil, nil, fmt.Errorf("reading module %q at branch %q HEAD: %w", moduleDir, branch, err)
```

Comment:
Yes that's correct. We don't want to skip it, we want to fail sync if any branch HEAD is broken.

We could retain the error and carry on, I'd have to add `Error() error` to `ModuleBranch` and document that it is a partial plan because an error may have occurred on a module branch. Let me know if you want me to go that direction.